src/neural/cuda/common_kernels.cu

```diff
@@ -566,6 +576,190 @@ void OutputInputTransform(int N, int C, int se_K, T* output, const T* input,
   ReportCUDAErrors(cudaGetLastError());
 }
 
+
+// N * C Tensors
+// performs softmax along the C dimension
+// Each thread processes one element
+// Sums are computed in shared memory
+// C threads per block, N blocks
+template <typename T>
+__global__ void softmax_kernel(T* output, const T* input) {
+  int n = blockIdx.x;
+  int c = threadIdx.x;
+  int C = blockDim.x;
+  int index = n * C + c;
+
+  __shared__ float sum;
+  if (c == 0) sum = 0;
+  __syncthreads();
+
+  // softmax = tf.exp(logits) / tf.reduce_sum(tf.exp(logits), axis)
+
+  float x = (float)input[index];
+  float ex = exp(x);
+
+  // compute warp wide sums first
+  float val = warpReduce(ex);
+
+  // update shared memory sum across C dimension
+  if ((c & 0x1F) == 0) atomicAdd(&sum, val);
+
+  __syncthreads();
+
+  float op = ex / sum;
+
+  output[index] = (T) op;
+}
+
+template <typename T>
+void Softmax(int N, int C, T* output, const T* input, cudaStream_t stream) {
+  softmax_kernel<T><<<N, C, 0, stream>>>(output, input);
+  ReportCUDAErrors(cudaGetLastError());
+}
+
+// N * C Tensors
+// performs layer normalization along the C dimension
+// Each thread processes one element
+// Sums/variences are computed in shared memory
+// C threads per block, N blocks
+template <typename T>
+__global__ void layer_norm_kernel(T* output, const T* input, const T* skip,
+                                  const T* gammas, const T* betas, float ep) {
+  int n = blockIdx.x;
+  int c = threadIdx.x;
+  int C = blockDim.x;
+
+  __shared__ float sum;
+  if (c == 0) sum = 0;
+  __syncthreads();
+
+  int index = n * C + c;
+
+  // From: https://www.tensorflow.org/api_docs/python/tf/keras/layers/LayerNormalization
+  // mean_i = sum(x_i[j] for j in range(k)) / k
+  // var_i  = sum((x_i[j] - mean_i) ^ 2 for j in range(k)) / k
+  // x_i_normalized = (x_i - mean_i) / sqrt(var_i + epsilon)
+  // output_i = x_i_normalized * gamma + beta
+
+  float x = (float)input[index];
+  if (skip) x += (float)skip[index];
+
+  float s = warpReduce(x);
+  if ((c & 0x1F) == 0) atomicAdd(&sum, s);
+
+  __syncthreads();
+
+  float mean = sum / C;
+  float d = x - mean;
+  float d_sq = d * d;
+
+  s = warpReduce(d_sq);
+  if ((c & 0x1F) == 0) atomicAdd(&sum, s);
```

Comment:
I think you need a reset of sum and a syncThreads after calculating the mean. Otherwise sum / C will be mean+variance. If mean and variance are similar scales you could just subtract the mean off the total to recover the variance - but I don't know how safe that is numeric stability wise in our use cases.

Reply:
yeah, very stupid of me to miss this! Better to keep two variables than adding more sync's.